exmg-paper-token-input.ts

```diff
@@ -0,0 +1,546 @@
+
+    if (this.autoValidate && !isInsideClick && this.previousClickWasInside) {
+      this.validate();
+    }
+
+    this.previousClickWasInside = isInsideClick;
+  }
+
+  private onIronInputKeyDown(e: KeyboardEvent): void {
+    switch (e.code || e.keyCode) {
+      case BACKSPACE:
+      case 'Backspace':
+        this.listBoxNode!.selectIndex(this.indexOfValue(this.selectedValues[this.selectedValues.length - 1]));
+        this.requestUpdate();
+        this.focusInputValue();
+        break;
+      case ARROWDOWN:
+      case 'ArrowDown':
+        this.menuElement!.open();
+        this.listBoxNode!.focus();
+        break;
+      case ESCAPE:
+      case 'Esc':
+        break;
+      default:
+        this.menuElement!.open();
+        afterNextRender(this, _ => this.focusInputValue());
+        break;
+    }
+  }
+
+  private onIronInputValueChanged(e: Event): void {
+    this.inputValue = (<HTMLInputElement>e.target).value || '';
+    this.inputValueNode!.style.width = `${this.inputWidthHelperNode!.offsetWidth + 10}px`;
+    this.filterItems();
+  }
+
+  private onPaperMenuVisibilityChanged(e: CustomEvent): void {
+    this.opened = e.detail.value;
+  }
+
+  private onInputContainerButtonTap(value: SelectedValue): () => void {
+    return () => {
+      this.selectedValues.splice(this.selectedValues.indexOf(value), 1);
+      this.focusInputValue();
+    };
+  }
+
+  private onInputContainerTap(e: Event): void {
+    e.preventDefault();
+    this.menuElement!.open();
+    afterNextRender(this, _ => this.focusInputValue());
+  }
+
+  private onInputFocusChanged(e: CustomEvent): void {
+    this.inputFocused = e.detail.value;
+  }
+
+  private onPaperListBoxItemSelect(e: CustomEvent): void {
+    if (this.maxTokens && this.selectedValues.length >= this.maxTokens) {
+      e.stopPropagation();
+    } else {
+      const value = this.getPaperItemValue(e.detail.item);
+
+      if (this.selectedValues.indexOf(value) === -1) {
+        this.selectedValues.push(value);
+        this.emitItemSelectEvent(value, e.detail.item);
+        this.resetInput();
+      }
+    }
+
+    this.requestUpdate();
+  }
+
+  private onPaperListBoxItemDeselect(e: CustomEvent): void {
+    const value = this.getPaperItemValue(e.detail.item);
+
+    if (this.selectedValues.indexOf(value) !== -1) {
+      this.selectedValues.splice(this.selectedValues.indexOf(value), 1);
+      this.emitItemDeselectEvent(value, e.detail.item);
+    }
+
+    this.resetInput();
+  }
+
+  /** END OF EVENT HANDLERS */
+
+  /** HELPERS */
+
+  private emitItemSelectEvent(value: SelectedValue, item: HTMLElement): void {
+    this.dispatchEvent(new CustomEvent('exmg-token-input-select', {
+      detail: {
+        value,
+        item,
+      },
+      bubbles: true,
+      composed: true,
+    }));
+  }
+
+  private emitItemDeselectEvent(value: SelectedValue, item: HTMLElement): void {
+    this.dispatchEvent(new CustomEvent('exmg-token-input-deselect', {
+      detail: {
+        value,
```

Comment:
Seems that detail is not necessary or at least value item is undefined ?

Reply:
no need